New Onboarding: Localize help contact url in Step-by-step launch flow

lsl opened the pull request:
Currently blocked by: https://github.com/Automattic/wp-calypso/issues/48033 - we need to wait until the logged out localized url is accessible before we localize this url. Logged in users already receive a localized experience.

#### Changes proposed in this Pull Request

* Wraps https://wordpress.com/help/contact link with `localizeUrl` in the etk launch flow.

#### Testing instructions

* Create a new site at /new
* Go through the launch flow from the editor and check the link is localized here: 
![Screen Shot on 2020-12-04 at 10:20:36](https://user-images.githubusercontent.com/1500769/101089723-6ff04900-361a-11eb-9b6b-35c6d3b27223.png)

Partial #37665

lsl committed:
Localize help contact url in launch flow

p-jackson reviewed (approved):
I was able to test it here in the launch flow:
![Screen Shot on 2020-12-04 at 10:20:36](https://user-images.githubusercontent.com/1500769/101089723-6ff04900-361a-11eb-9b6b-35c6d3b27223.png)

LGTM

lsl commented:
Thanks, tested, the change itself is LGTM too: `localizeUrl` is now being applied in ETK.

There are some blocking issues with this url localization so shifting back to draft and adding do not merge.

`Logged in: https://wordpress.com/help/contact > 200` :heavy_check_mark:  
`Logged out: https://wordpress.com/help/contact > 302 https://wordpress.com/support/contact/ > 200` :heavy_check_mark: 
`Logged in: https://es.wordpress.com/help/contact > 302 https://wordpress.com/me/help > 404` :negative_squared_cross_mark: 
`Logged out: https://es.wordpress.com/help/contact > 302 https://wordpress.com/support/ > 200` :negative_squared_cross_mark: not localized, should be `/es/support/contact` but that page doesn't exist.

To fix this we need a few pieces put together:

- Add a `localizeUrl` rule for `/help/contact` to localize to `/help/contact/es`.
- Add a calypso route for `/help/contact/:locale` to redirect to `/help/contact` if logged in. Or /es/support/contact if logged out.
- Add localized versions of /:locale/support/contact. This is a special theme page that probably just needs to be enabled and configured for each locale depending on whether we have localized support available or not.

akirk commented on apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/src/launch-steps/final-step/index.tsx:
```diff
@@ -163,7 +166,11 @@ const FinalStep: React.FunctionComponent< LaunchStepProps > = ( { onNextStep, on
-										<Button isLink href="https://wordpress.com/help/contact" target="_blank">
+										<Button
+											isLink
+											href={ localizeUrl( 'https://wordpress.com/help/contact' ) }
```
After #48645 you can tell `localizeUrl` if we are logged-in or logged-out:
```suggestion
											href={ localizeUrl( 'https://wordpress.com/help/contact', null, window.currentUser ) }
```

lsl committed:
Update apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/src/launch-steps/final-step/index.tsx

Co-authored-by: Alex Kirk <akirk@users.noreply.github.com>

simison commented:
@razvanpapadopol could this be on Luna's project backlog as well? Looks like pretty much ready to go. Thanks!

razvanpapadopol commented on apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/src/launch-steps/final-step/index.tsx:
```diff
@@ -200,7 +203,11 @@ const FinalStep: React.FunctionComponent< LaunchStepProps > = ( { onNextStep, on
-										<Button isLink href="https://wordpress.com/help/contact" target="_blank">
+										<Button
+											isLink
+											href={ localizeUrl( 'https://wordpress.com/help/contact', null, window.currentUser ) }
```
There is a TypeScript error because of this line:
> Argument of type 'null' is not assignable to parameter of type 'string | undefined'.

Then,`window.currentUser` seem to be always `undefined` in this context but we know that a user needs to be logged in to be able to launch their site, right? 

After checking the `useLocalizeUrl` callback arguments at https://github.com/Automattic/wp-calypso/blob/trunk/packages/i18n-utils/src/localize-url.tsx#L153 the correct function call would be:
```
localizeUrl( 'https://wordpress.com/help/contact', locale, true )
``` 

But `localizeUrl` at https://github.com/Automattic/wp-calypso/blob/trunk/packages/i18n-utils/src/localize-url.tsx#L104 defaults last argument (` isLoggedIn`) to true so we can just omit that.
```suggestion
											href={ localizeUrl( 'https://wordpress.com/help/contact', locale ) }
```

StefanNieuwenhuis committed:
Update props passed to localizeUrl

`localizeUrl` expects an URL, locale and isLoggedIn boolean to be passed. Since `isLoggedIn` defaults to `true` (since a user has to be logged in to launch a site) we only have to pass the correct locale. This commit fixes this.

Co-authored-by: Răzvan Papadopol <razvan.papadopol@automattic.com>

akirk commented on apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/src/launch-steps/final-step/index.tsx:
```diff
@@ -200,7 +203,11 @@ const FinalStep: React.FunctionComponent< LaunchStepProps > = ( { onNextStep, on
-										<Button isLink href="https://wordpress.com/help/contact" target="_blank">
+										<Button
+											isLink
+											href={ localizeUrl( 'https://wordpress.com/help/contact', locale ) }
```
#48033 mentions that the original change in this PR (with the third argument) was necessary because (apparently) depending on whether a new user creates a new site, we need different support URLs. A logged-in user can use the Calypso support form while a logged-out user needs to go to the support site. Therefore, it's important here to pass in the state of the user.
```suggestion
											href={ localizeUrl( 'https://wordpress.com/help/contact', locale, undefined !== window.currentUser ) }
```

razvanpapadopol replied:
> A logged-in user can use the Calypso support form while a logged-out user needs to go to the support site. 

While the logged-out user scenario applies for some landing pages and signup flows in Calypso, this part of the application is loaded only for an existing unlaunched wpcom site (simple or atomic) with ETK plugin and accessible only by logged-in users from the block editor.

Also, please note that `window.currentUser` is `undefined` in the iframed editor where this code runs.

akirk replied:
Ok, just wanted to make sure because I built #48645 to accommodate the logged in/logged out scenario that was requested for this PR. If that assumption is not valid, then all good.

akirk reviewed (changes requested):
I believe the currentUser needs to be taken into account.
